`cfg_select!` macro

folkertdev opened the pull request:
Stabilization report: https://github.com/rust-lang/rust/pull/149783

folkertdev committed:
`cfg_select!` macro

folkertdev committed:
more detail

folkertdev commented on src/conditional-compilation.md:
````diff
@@ -466,11 +466,75 @@ let machine_kind = if cfg!(unix) {
+```
+
+r[cfg.cfg_select.general]
+The built-in `cfg_select` macro expands to the `TokenTree` on the right-hand side of the first configuration predicate that evaluates to `true`.
````
is it clear enough from the grammer and the language here that the `{ /* ... */ }` are dropped?

folkertdev reviewed (commented):
I think all the raw information is here now, but it probably needs some editing/clarification.

Also should the lint on branches that come after a wildcard be mentioned?

ehuss reviewed (commented):
> Also should the lint on branches that come after a wildcard be mentioned?

As for lints, we normally don't document those. However, in this situation, it seems like it would be fine to have a `> [!NOTE]` block that mentions the lint (in whichever way https://github.com/rust-lang/rust/pull/149960 ends up being).

traviscross committed:
Add more correct grammar for `cfg_select!`

Macros can be renamed when imported and can delimit their bodies with
different tokens, so we don't want to document the outer part of the
call; we just want to document the input to the macro.

Doing this for `cfg_select!` is a bit tricky since an expression with
a block doesn't need to be followed by a comma while an expression
without a block does unless it's the last one.  We don't want to
handle this the way that the `match` grammar does, currently, since we
want to move toward finite lookahead.  So, instead, we handle this
with right recursion in `CfgSelectArms`.

Unlike `match`, outer attributes aren't allowed on the RHS
expressions.  To handle this, we need to refactor
`ExpressionWithoutBlock` and `ExpressionWithBlock`.

We also document that the outer braces are removed during expansion
when the payload is a block expression.

(We called it the arm "payload" after running out of other options.)

ehuss commented on src/conditional-compilation.md:
```diff
@@ -466,17 +466,85 @@ let machine_kind = if cfg!(unix) {
+    CfgSelectConfigurationPredicate `=>`
+    (
+        ExpressionWithBlockNoAttrs `,`? CfgSelectArms?
+      | ExpressionWithoutBlockNoAttrs ( `,` CfgSelectArms? )?
```
I think this needs to include `{` TokenTree `}`. Otherwise, this wouldn't match other non-expression things like patterns, types, etc.

ehuss commented on src/conditional-compilation.md:
```diff
@@ -466,11 +466,75 @@ let machine_kind = if cfg!(unix) {
+- impl items
+- trait impl items
+- trait items
+- foreign items
```
Also, did you intend to exclude patterns and types in this list?

folkertdev replied:
I suspect that the list was accurate at the time, but it no longer is and should now include patterns and types.

folkertdev committed:
Update src/conditional-compilation.md

Co-authored-by: Eric Huss <eric@huss.org>

tshepang commented on src/conditional-compilation.md:
````diff
@@ -466,11 +466,78 @@ let machine_kind = if cfg!(unix) {
+    ConfigurationPredicate | `_`
+```
+
+r[cfg.cfg_select.general]
````
@ehuss should this be in dev-guide

ehuss committed:
Rework the cfg_select intro

The rule names for intros should be `.intro`. The intro itself should
not be specifying the behavior. Also, follow the style where there is an
introduction, an example, and then the syntax.

ehuss committed:
Add specific rules for cfg_select behavior

ehuss committed:
Simplify wording of cfg.cfg_select.wildcard

ehuss committed:
Remove cfg.cfg_select.positions

Don't repeat the list of valid macro positions (which is already not
correct here since it is missing some). For the most part, it is assumed
that macro calls can appear in all valid macro invocation positions. If
exceptions are ever added in the future, then they can be called out.

ehuss committed:
Simplify wording of cfg.cfg_select.fallthrough

ehuss committed:
Improve wording of cfg.cfg_select.well-formed

ehuss committed:
Fix sorting of cfg_select! link definition

ehuss reviewed (approved):
Thanks! Merging now that https://github.com/rust-lang/rust/pull/149783 is on nightly.


*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/reference/2103/70c99cb152e25d6d134d952be868a8c53c4c894f..5c1d0c5a9b616348d036f780d806ede5ecbd4f3c)*
